STACK-21374 Backup and restore ClickHouse (#1471)
StackVista/stackstate-docs: setup/data-management/backup_restore/kubernetes_backup.md

- `victoria-metrics-0` - 25 minutes past the hour
- `victoria-metrics-1` - 35 minutes past the hour

> The backup schedule can be configured using the Helm value `victoria-metrics-0.backup.scheduled.schedule` according [cronexpr format](https://github.com/aptible/supercronic/tree/master/cronexpr)

Reviewer: ```suggestion
The backup schedule can be configured using the Helm value `victoria-metrics-0.backup.scheduled.schedule` according to [cronexpr format](https://github.com/aptible/supercronic/tree/master/cronexpr)
```

## OpenTelemetry \(ClickHouse\)

> ClickHouse uses both incremental and full backups. By default, full backups are executed daily at 00:45 am, and incremental backups are performed every hour. Each backup creates a new directory, old backups (directories) are deleted automatically. All files located in a backup directory should be treated as a single whole and can only be moved, copied or deleted as a whole. We recommend to uses `clickhouse-backup` tool to manage backups. The tool is available on the `stackstate-clickhouse-shard0-0` Pod.

Reviewer: ```suggestion
ClickHouse uses both incremental and full backups. By default, full backups are executed daily at 00:45 am, and incremental backups are performed every hour. Each backup creates a new directory, old backups (directories) are deleted automatically. All files located in a backup directory should be treated as a single whole and can only be moved, copied or deleted as a whole. We recommend to use `clickhouse-backup` tool to manage backups. The tool is available on the `stackstate-clickhouse-shard0-0` Pod.
```

### Enable scheduled backups

Backups of the ClickHouse are disabled by default. To enabled scheduled ClickHouse backups, set the Helm value `clickhouse.backup.enabled` to `true`.

{% hint style="warning" %}
ClickHouse backups requires to [enable backups](#enable-backups).
{% endhint %}

### Enable restores

> Restore functionality of the ClickHouse are disabled by default. To enabled restore functionality of the ClickHouse, set the Helm value `clickhouse.restore.enabled` to `true`.

Reviewer: ```suggestion
Restore functionality of the ClickHouse is disabled by default. To enable restore functionality of the ClickHouse, set the Helm value `clickhouse.restore.enabled` to `true`.
```

{% hint style="warning" %}
ClickHouse restore functionality requires to [enable backups](#enable-backups).
{% endhint %}

### Backup schedule

By default, the ClickHouse backups are created:
- Full Backup - at 00:45 every day
- Incremental Backup - 45 minutes past the hour (from 3 am to 12 am)

{% hint style="warning" %}
Backups struggle with parallel execution. If a second backup starts before the first one completes, it will disrupt the first backup. Therefore, it's crucial to avoid parallel execution. For instance, the first incremental backup should be executed three hours after the full one.
{% endhint %}

> The backup schedule can be configured using the Helm value `clickhouse.backup.scheduled.full_schedule` and `clickhouse.backup.scheduled.incremental_schedule` according [cronexpr format](https://github.com/aptible/supercronic/tree/master/cronexpr)

Reviewer: ```suggestion
The backup schedule can be configured using the Helm value `clickhouse.backup.scheduled.full_schedule` and `clickhouse.backup.scheduled.incremental_schedule` according to [cronexpr format](https://github.com/aptible/supercronic/tree/master/cronexpr)
```

```

The output should look like this:

```bash
full_2024-06-17T18-50-00          34.41KiB   17/06/2024 18:50:00   remote                                      tar, regular
incremental_2024-06-17T18-51-00   7.29KiB    17/06/2024 18:51:00   remote   +full_2024-06-17T18-50-00          tar, regular
incremental_2024-06-17T18-54-00   7.29KiB    17/06/2024 18:54:00   remote   +incremental_2024-06-17T18-51-00   tar, regular
incremental_2024-06-17T18-57-00   7.29KiB    17/06/2024 18:57:00   remote   +incremental_2024-06-17T18-54-00   tar, regular
full_2024-06-17T19-00-00          26.41KiB   17/06/2024 19:00:00   remote                                      tar, regular
incremental_2024-06-17T19-00-00   6.52KiB    17/06/2024 19:00:00   remote   +incremental_2024-06-17T18-57-00   tar, regular
incremental_2024-06-17T19-03-00   25.37KiB   17/06/2024 19:03:00   remote   +incremental_2024-06-17T19-00-00   tar, regular
incremental_2024-06-17T19-06-00   7.29KiB    17/06/2024 19:06:00   remote   +incremental_2024-06-17T19-03-00   tar, regular
```
where is printed:
- name, the name started with `full_` - it is a full backup, `incremental_` - it is an incremental backup 
- size,
- creation date,

> - `remote` - a backup is upload to a remote storage like S3

Reviewer: ```suggestion
- `remote` - a backup is uploaded to a remote storage like S3
```

2024/06/17 19:14:19.509574  info done                      backup=incremental_2024-06-17T19-06-00 duration=66ms operation=restore_data
2024/06/17 19:14:19.509591  info done                      backup=incremental_2024-06-17T19-06-00 duration=167ms operation=restore version=2.5.13
2024/06/17 19:14:19.509684  info clickhouse connection closed logger=clickhouse
Data restored
```

{% hint style="info" %}
**Error: error can't create table …. code: 57, message: Directory for table data store/…/ already exists**

ClickHouse does not permanently delete tables when the `DROP DATABASE/TABLE ...` command is executed. Instead, the database is marked as deleted and will be permanently removed after 8 minutes. This delay provides additional time to undo the operation. More details can be found at [UNDROP TABLE](https://clickhouse.com/docs/en/sql-reference/statements/undrop). If you attempt to restore data during this period, it will fail and produce the aforementioned error. Available solutions:
- wait 8 minutes (check table `select * from system.dropped_tables;` )

> - configure `database_atomic_delay_before_drop_table_sec`

Reviewer: Regarding these notes:
```
(check table `select * from system.dropped_tables;` )
- configure `database_atomic_delay_before_drop_table_sec`
```
I would say we either have to explain how to do that or remove them...
